### PR TITLE
JAMES-2838 Bucket support for the CassandraBlobStore

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Comment by chibenwa** [comment]
```
Progress (17): 1.0/2.7 MB | 581 kB | 0.5/4.0 MB | 0.4/2.4 MB | 348/500 kB | 0.3/1.3 MB | 258/324 kB | 188/252 kB | 0.3/1.0 MB[ERROR] Failed to execute goal on project james-server-util: Could not resolve dependencies for project org.apache.james:james-server-util:bundle:3.4.0-SNAPSHOT: Failed to collect dependencies at com.github.steveash.guavate:guavate:jar:1.0.0 -> com.google.guava:guava:jar:25.1-jre -> com.google.errorprone:error_prone_annotations:jar:2.1.3: Failed to read artifact descriptor for com.google.errorprone:error_prone_annotations:jar:2.1.3: Could not transfer artifact com.google.errorprone:error_prone_parent:pom:2.1.3 from/to central (https://repo.maven.apache.org/maven2): /root/.m2/repository/com/google/errorprone/error_prone_parent/2.1.3/error_prone_parent-2.1.3.pom.part (No such file or directory) -> [Help 1]
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3]  | 163/197 kB [ERROR] 
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] |[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3]  [ERROR] Re-run Maven using the -X switch to enable full debug logging.
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] 7[ERROR] 
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [ERROR] For more information about the errors and possible solutions, please read the following articles:
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] 0/281 kB | 44[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3]  kB [ERROR] 
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [ERROR] After correcting the problems, you can resume the build with the command
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] | 61/89 kB | 16/108 kB | 36/215 kB | 15 kB | 2.2 kB
[ERROR]   mvn <goals> -rf :james-server-util
```

test this please

**Comment by trantienduchn** [comment]
```
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [INFO] /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java: /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java uses unchecked or unsafe operations.
[b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [INFO] /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java: Recompile with -Xlint:unchecked for details.
```

**Comment by Arsnael** [comment]
> ```
> [b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [INFO] /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java: /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java uses unchecked or unsafe operations.
> [b2969d5f219f54c101142a4bd10d3c8b2379e8b3] [INFO] /james-project/event-sourcing/event-store-cassandra/src/main/java/org/apache/james/eventsourcing/eventstore/cassandra/JsonEventSerializer.java: Recompile with -Xlint:unchecked for details.
> ```

Weird, that seems coming from last merged PRs on event serializer, not this one, but on local I can compile the latest master branch... Perhaps that build was running between two merges?

test this please

**Comment by Arsnael** [comment]
Overall I don't have anything to say about the code, but more global questions... 
* my guess is the default bucket would be the one actually in use, like the current month we are? Just cannot figure out in the code how do we switch to the new default bucket when reaching a new month? (might have missed that point on my review)
* are the proper guice bindings for this gonna be handled in the cleanup task?

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobModule.java [hunk 1]
```diff
@@ -19,35 +19,53 @@
 
 package org.apache.james.blob.cassandra;
 
-import static org.apache.james.blob.cassandra.BlobTables.BlobParts.CHUNK_NUMBER;
-import static org.apache.james.blob.cassandra.BlobTables.BlobParts.DATA;
-import static org.apache.james.blob.cassandra.BlobTables.BlobTable.ID;
-import static org.apache.james.blob.cassandra.BlobTables.BlobTable.NUMBER_OF_CHUNK;
-import static org.apache.james.blob.cassandra.BlobTables.BlobTable.TABLE_NAME;
+import static org.apache.james.blob.cassandra.BlobTables.DefaultBucketBlobParts.DATA;
```
Comment (mbaechler) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobModule.java]: we should not use static import now that we have several similar tables, don't you think?

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreTest.java [hunk 2]
```diff
@@ -102,6 +103,12 @@ public void deleteBucketShouldThrowWhenNullBucketName() {
 
     }
 
+    @Override
+    @Disabled("Concurrent read and delete can lead to partial reads (no transactions)")
```
Comment (mbaechler) [inline server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreTest.java]: Are you sure it can't be done?

**Comment by trantienduchn** [comment]
> my guess is the default bucket would be the one actually in use, like the current month we are? Just cannot figure out in the code how do we switch to the new default bucket when reaching a new month? (might have missed that point on my review) 

The switch is don in `CassandraBlobStore`, for example `CassandraBlobStore.saveBlobPartReference()`. We will store everything in the default bucket, except deleted messages should be stored in other buckets. For storing deleted messages, when switching to new buckets, we just use another `bucketId` as it is a part of the Primary Key in cassandra `blobPartsInBucket` table

File: server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreTest.java (hunk 2 above)
Comment (chibenwa) [reply server/blob/blob-cassandra/src/test/java/org/apache/james/blob/cassandra/CassandraBlobStoreTest.java]: Can be done: I was looking at the wrong root-cause. The root cause is the fact that PipedStream do not propagate exception. Extending them allows to solve this.

**Comment by chibenwa** [comment]
> my guess is the default bucket would be the one actually in use, like the current month we are? Just cannot figure out in the code how do we switch to the new default bucket when reaching a new month? (might have missed that point on my review) And also the migration of data between the default bucket and bucket tables when this happen? Or this happens on a higher level I guess?

You are looking with a functional eye what is a technical thing. And getting confused. (Why are we speaking of vault here?)

The wault do not rely on the default bucket, and Cassandra always returns BucketName.DEFAULT hence I don't undestand your question.

> are the proper guice bindings for this gonna be handled in the cleanup task?

For sure: Cassandra cleanup is done for every created table. I bet your are mixing things with the isolation strategy chosen for ObjectStorage (not the same thing!)

**Comment by chibenwa** [comment]
```
[INFO] Apache James :: Server :: Data :: JMAP :: Cassandra persistance FAILURE [ 38.804 s]

AccessTokenRepositoryTest.validTokenMustBeRetrieved:45 » InvalidAccessToken 78...
```

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by Arsnael** [comment]
```
[2687f0adf1954fc81b7941b452c5ec9a5351da37] [ERROR] Errors: 
[2687f0adf1954fc81b7941b452c5ec9a5351da37] [ERROR]   DeleteBlobStoreContract.readBytesShouldNotReadPartiallyWhenDeletingConcurrentlyBigBlob:199 » Execution
```

**Comment by Arsnael** [comment]
After trying to check the logs deeper (which was not easy, thanks to heavy logs), I found this : 
```
[ERROR] Tests run: 46, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 44.489 s <<< FAILURE! - in org.apache.james.blob.objectstorage.ObjectStorageBlobsDAOAWSPrefixTest
[ERROR] readBytesShouldNotReadPartiallyWhenDeletingConcurrentlyBigBlob  Time elapsed: 2.665 s  <<< ERROR!
java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.RuntimeException: Should not read partial blob when an other thread is deleting it. Size : 10444584
	at org.apache.james.blob.api.DeleteBlobStoreContract.readBytesShouldNotReadPartiallyWhenDeletingConcurrentlyBigBlob(DeleteBlobStoreContract.java:199)
Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Should not read partial blob when an other thread is deleting it. Size : 10444584
Caused by: java.lang.RuntimeException: Should not read partial blob when an other thread is deleting it. Size : 10444584
	at org.apache.james.blob.api.DeleteBlobStoreContract.lambda$readBytesShouldNotReadPartiallyWhenDeletingConcurrentlyBigBlob$6(DeleteBlobStoreContract.java:189)
```
Which makes no sense as this test is disabled in `ObjectStorageBlobsDAOAWSPrefixTest`. Any idea?

**Comment by chibenwa** [comment]
:+1: with your fixes

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+    @Override
+    public BucketName getDefaultBucketName() {
+        return BucketName.DEFAULT;
+    }
+
+    private Flux<byte[]> readBlobParts(BucketName bucketName, BlobId blobId) {
+        Integer rowCount = selectRowCount(bucketName, blobId)
+            .publishOn(Schedulers.elastic())
+            .switchIfEmpty(Mono.error(
```
Comment (blackheaven) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBlobStore.java]: I prefer that you use `single()`.

File: server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBucketDAO.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    Flux<Pair<BucketName, BlobId>> listAll() {
+        return cassandraAsyncExecutor.executeRows(listAll.bind())
+            .map(row -> Pair.of(BucketName.of(row.getString(BUCKET)), blobIdFactory.from(row.getString(ID))));
+    }
+
+    private byte[] rowToData(Row row) {
+        byte[] data = new byte[row.getBytes(BucketBlobParts.DATA).remaining()];
+        row.getBytes(BucketBlobParts.DATA).get(data);
```
Comment (blackheaven) [inline server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/CassandraBucketDAO.java]: no checks?

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**